sms_alternative_provider/models/sms_sms.py

```diff
@@ -0,0 +1,41 @@
+# Copyright 2024 Hunki Enterprises BV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl-3.0)
+
+from odoo import fields, models
+
+
+class SmsSms(models.Model):
+    _inherit = "sms.sms"
+
+    sms_gateway_id = fields.Many2one("ir.sms.gateway", string="SMS gateway used")
+
+    def _postprocess_iap_sent_sms(
+        self, iap_results, failure_reason=None, unlink_failed=False, unlink_sent=True
+    ):
+        """
+        Group results by provider, let super handle actual iap results, divert others
+        to their ir.sms.gateway handler
+        """
+        IrSmsGateway = self.env["ir.sms.gateway"]
+        provider2results = {}
+        for result in iap_results:
+            provider2results.setdefault(
+                IrSmsGateway.browse(result.get("sms_gateway_id", [])), []
+            ).append(result)
+
+        actual_iap_results = []
+        for provider, results in provider2results.items():
+            if not provider or provider.gateway_type == "iap":
+                actual_iap_results.extend(results)
+            else:
+                provider._handle_results(
+                    [], results, unlink_failed=unlink_failed, unlink_sent=unlink_sent
```

Comment:
Hey, the message dicts are not passed here,

Reply:
there are none at this point, because this function is called in `sms.sms#_send`, which operates on `sms.sms` records, and those you can access via `results[i]['id']`

actually I don't understand any more why you requested having the message dicts passed here in the first place. If a provider wants to do provider-specific stuff, why wouldn't it do that in `_send_$provider` after sending? And note that whatever a provider wants to write on an sms.sms record, it can just add to its result dictionaries.